File: src/lib/geoip.ts
/**
 * GeoIP lookup using ip-api.com (free, no key needed, 45 req/min).
 * Results are cached in-memory to avoid repeated lookups for the same IP.
 */

interface GeoResult {
  country: string | null;
  city: string | null;
  lat: number | null;
  lon: number | null;
}

const cache = new Map<string, GeoResult>();
const EMPTY: GeoResult = { country: null, city: null, lat: null, lon: null };

// Private/reserved IPs — skip lookup
function isPrivateIp(ip: string): boolean {
  return /^(10\.|172\.(1[6-9]|2\d|3[01])\.|192\.168\.|127\.|::1|fc|fd|fe80)/.test(ip);
}

export async function geoLookup(ip: string): Promise<GeoResult> {
  if (!ip || isPrivateIp(ip)) return EMPTY;

  const cached = cache.get(ip);
  if (cached) return cached;

  try {
    const res = await fetch(`http://ip-api.com/json/${ip}?fields=status,country,city,lat,lon`, {
      signal: AbortSignal.timeout(3000),
    });
    const data = await res.json();

    if (data.status === "success") {
      const result: GeoResult = {
        country: data.country ?? null,
        city: data.city ?? null,
        lat: data.lat ?? null,
        lon: data.lon ?? null,
      };
      cache.set(ip, result);
      return result;
    }
  } catch {
    // Timeout or network error — don't block event ingestion
  }

  cache.set(ip, EMPTY);
  return EMPTY;
}

/** Batch lookup — ip-api.com supports batch of up to 100 IPs */
export async function geoBatchLookup(ips: string[]): Promise<Map<string, GeoResult>> {
  const results = new Map<string, GeoResult>();
  const toFetch: string[] = [];

  for (const ip of ips) {
    if (!ip || isPrivateIp(ip)) {
      results.set(ip, EMPTY);
    } else if (cache.has(ip)) {
      results.set(ip, cache.get(ip)!);
    } else {
      toFetch.push(ip);
    }
  }

  if (toFetch.length > 0) {
    try {
      const res = await fetch("http://ip-api.com/batch?fields=status,query,country,city,lat,lon", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(toFetch.slice(0, 100)),
        signal: AbortSignal.timeout(5000),
      });
      const data: Array<{ status: string; query: string; country?: string; city?: string; lat?: number; lon?: number }> = await res.json();

      for (const d of data) {
        const result: GeoResult = d.status === "success"
          ? { country: d.country ?? null, city: d.city ?? null, lat: d.lat ?? null, lon: d.lon ?? null }
          : EMPTY;
        cache.set(d.query, result);
        results.set(d.query, result);
      }
    } catch {
      for (const ip of toFetch) {
        results.set(ip, EMPTY);
      }
    }
  }

  return results;
}

File: worker/internal/geoip/geoip.go
package geoip

import (
	"bytes"
	"encoding/json"
	"net"
	"net/http"
	"sync"
	"time"
)

type Result struct {
	Country string  `json:"country"`
	City    string  `json:"city"`
	Lat     float64 `json:"lat"`
	Lon     float64 `json:"lon"`
}

var (
	cache   sync.Map
	client  = &http.Client{Timeout: 5 * time.Second}
	empty   = Result{}
)

func isPrivate(ip string) bool {
	p := net.ParseIP(ip)
	if p == nil {
		return true
	}
	return p.IsLoopback() || p.IsPrivate() || p.IsLinkLocalUnicast()
}

// BatchLookup resolves up to 100 IPs via ip-api.com batch endpoint.
func BatchLookup(ips []string) map[string]Result {
	results := make(map[string]Result, len(ips))
	var toFetch []string

	for _, ip := range ips {
		if isPrivate(ip) {
			results[ip] = empty
			continue
		}
		if v, ok := cache.Load(ip); ok {
			results[ip] = v.(Result)
			continue
		}
		toFetch = append(toFetch, ip)
	}

	if len(toFetch) == 0 {
		return results
	}

	// ip-api.com batch: max 100
	for i := 0; i < len(toFetch); i += 100 {
		end := i + 100
		if end > len(toFetch) {
			end = len(toFetch)
		}
		batch := toFetch[i:end]

		body, _ := json.Marshal(batch)
		resp, err := client.Post(
			"http://ip-api.com/batch?fields=status,query,country,city,lat,lon",
			"application/json",
			bytes.NewReader(body),
		)
		if err != nil {
			for _, ip := range batch {
				results[ip] = empty
			}
			continue
		}

		var data []struct {
			Status  string  `json:"status"`
			Query   string  `json:"query"`
			Country string  `json:"country"`
			City    string  `json:"city"`
			Lat     float64 `json:"lat"`
			Lon     float64 `json:"lon"`
		}
		json.NewDecoder(resp.Body).Decode(&data)
		resp.Body.Close()

		for _, d := range data {
			r := empty
			if d.Status == "success" {
				r = Result{Country: d.Country, City: d.City, Lat: d.Lat, Lon: d.Lon}
			}
			cache.Store(d.Query, r)
			results[d.Query] = r
		}
	}

	return results
}
